thingif/src/main/java/com/kii/thingif/ThingIFAPI.java

```diff
@@ -2170,7 +2173,51 @@ public String getFirmwareVersion() throws ThingIFException {
             @NonNull Aggregation aggregation,
             @NonNull Class<S> targetStateClass,
             @NonNull Class<T> valueClass) throws ThingIFException {
-        //TODO: // FIXME: 12/21/16 implement the logic
-        return new ArrayList<>();
+        if (this.target == null) {
+            throw new IllegalStateException("Can not perform this action before onboarding");
+        }
+        if (!this.stateTypes.containsKey(groupedQuery.getAlias())) {
+            throw new UnregisteredAliasException(groupedQuery.getAlias(), false);
+        }
+
+        Gson localGson = new GsonBuilder()
+                .registerTypeAdapter(GroupedHistoryStatesQuery.class,
+                        new GroupedHistoryStatesQueryAdapter(aggregation))
+                .registerTypeAdapter(AggregatedResult.class,
+                        new AggregatedResultAdapter(targetStateClass, valueClass))
```

Comment:
make sure there is not uncheck warning. Fixed by adding `<>` to AggregatedResultAdapter
```java
new AggregatedResultAdapter<>(targetStateClass, valueClass))
```